### DISCUSSION TITLE
thread/thread-bring-some-problem

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#247] [Announcements]
# thread/thread-bring-some-problem

多线程技术有很多好处，比如说多线程可以充分利用多核 CPU 的计算能力，但如果使用不慎，就会出现『线程安全问题』『活跃性问题』『性能问题』等等。

https://javabetter.cn/thread/thread-bring-some-problem.html

**Comment by hengxingstu** [comment]
原文有误，四个操作中，第三个int j = i;不具备原子性，因为这一步需要先读取i的值，再对进行赋值。

**Reply by hengxingstu** [reply]
对 j 进行赋值

**Reply by Msxiaoxu** [reply]
其实可能是写错了，作者先说了只有1是原子操作。

有些小伙伴可能认为这些都是原子操作，其实只有操作 1 是原子操作。

操作 1：这是原子操作，因为它是一个单一的、不可分割的步骤。
操作 2：这不是原子操作。这实际上是一个 "read-modify-write" 操作，它包括了读取 i 的值，增加 i，然后写回 i。
操作 3：这是原子操作，因为它是一个单一的、不可分割的步骤。
操作 4：这不是原子操作。和 i++ 一样，这也是一个 "read-modify-write" 操作。

**Reply by hengxingstu** [reply]
所以确实是有错误的咯，应该改成如下内容：

>操作 1：这是原子操作，因为它是一个单一的、不可分割的步骤。
操作 2：这不是原子操作。这实际上是一个 "read-modify-write" 操作，它包括了读取 i 的值，增加 i，然后写回 i。
操作 3：这不是原子操作，需要先读取 i 的值，再对 j 进行赋值。
操作 4：这不是原子操作。和 i++ 一样，这也是一个 "read-modify-write" 操作。

**Comment by zhuhui001** [comment]
文中提到了可见性可以用volatile关键字或者syn/lock，那原子性怎么保证

**Reply by hengxingstu** [reply]
我的理解是这样：
首先，原文中对原子性的描述是这样：
> 原子性：即一个操作或者多个操作，要么全部执行并且执行的过程不会被任何因素打断，要么就都不执行。
也就是说，实现原子性，就是保证同一时间只有一个程序执行同步代码块。这一点可以通过系统的重量级锁来实现，或者可以通过CAS操作来实现。
这些在后文有介绍，不知道我的理解是否完全正确，若二哥看到我的理解有错误的地方，请不吝指教

**Comment by cwj2022** [comment]
怎么解决多线程产生的问题呢？这里好像没写

**Reply by hengxingstu** [reply]
后面的文章有介绍
